src/schema/schema_alter.c

```diff
@@ -0,0 +1,190 @@
+/*-
+ * Copyright (c) 2014-2016 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __alter_file --
+ *	Alter a file.
+ */
+static int
+__alter_file(
+    WT_SESSION_IMPL *session, const char *uri, const char *newcfg[])
+{
+	WT_DECL_RET;
+	const char *cfg[4], *filename;
+	char *config, *newconfig;
+
+	filename = uri;
+	newconfig = NULL;
+	if (!WT_PREFIX_SKIP(filename, "file:"))
+		return (__wt_unexpected_object_type(session, uri, "file:"));
+
+	/* Find the URI */
+	WT_RET(__wt_metadata_search(session, uri, &config));
+
+	WT_ASSERT(session, newcfg[0] != NULL);
+	/*
+	 * Start with the base configuration because collapse is like
+	 * a projection and if we are reading older metadata, it may not
+	 * have all the components.
+	 */
+	cfg[0] = WT_CONFIG_BASE(session, file_meta);
+	cfg[1] = config;
+	cfg[2] = newcfg[0];
+	cfg[3] = NULL;
+	WT_ERR(__wt_config_collapse(session, cfg, &newconfig));
+	/*
+	 * Only rewrite if there are changes.
+	 */
+	if (strcmp(config, newconfig) != 0)
+		WT_ERR(__wt_metadata_update(session, uri, newconfig));
+	else
+		WT_STAT_CONN_INCR(session, session_table_alter_skip);
+
+err:	__wt_free(session, config);
+	__wt_free(session, newconfig);
+	return (ret);
+}
+
+/*
+ * __alter_colgroup --
+ *	WT_SESSION::alter for a colgroup.
+ */
+static int
+__alter_colgroup(
+    WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
+{
+	WT_COLGROUP *colgroup;
+	WT_DECL_RET;
+
+	WT_ASSERT(session, F_ISSET(session, WT_SESSION_LOCKED_TABLE));
+
+	/* If we can get the colgroup, perform any potential alterations. */
+	if ((ret = __wt_schema_get_colgroup(
+	    session, uri, false, NULL, &colgroup)) == 0)
+		WT_TRET(__wt_schema_alter(session, colgroup->source, cfg));
+
+	return (ret);
+}
+
+/*
+ * __alter_index --
+ *	WT_SESSION::alter for an index.
+ */
+static int
+__alter_index(
+    WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
+{
+	WT_INDEX *idx;
+	WT_DECL_RET;
+
+	/* If we can get the index, perform any potential alterations. */
+	if ((ret = __wt_schema_get_index(
+	    session, uri, false, NULL, &idx)) == 0)
+		WT_TRET(__wt_schema_alter(session, idx->source, cfg));
+
+	return (ret);
+}
+
+/*
+ * __alter_table --
+ *	WT_SESSION::alter for a table.
+ */
+static int
+__alter_table(WT_SESSION_IMPL *session, const char *uri, const char *cfg[])
+{
+	WT_COLGROUP *colgroup;
+	WT_DECL_RET;
+	WT_TABLE *table;
+	const char *tblcfg[2], *name;
+	u_int i;
+	int ncolgroups;
+
+	name = uri;
+	(void)WT_PREFIX_SKIP(name, "table:");
+
+	WT_RET(__wt_schema_get_table(
+	    session, name, strlen(name), true, &table));
+
+	/*
+	 * Get the original configuration for the table to see if it is
+	 * using the default column group.
+	 */
+	tblcfg[0] = table->config;
+	tblcfg[1] = NULL;
+	WT_ERR(__wt_table_colgroups_config(session, tblcfg, &ncolgroups));
```

Comment:
I believe this code can just check `table->ncolgroups == 0` because that is resolved by `__wt_schema_get_table`.  Is there some more complicated case I'm missing?